v1/ast/term.go

```diff
@@ -1017,6 +1017,30 @@ func (ref Ref) Copy() Ref {
 	return termSliceCopy(ref)
 }
 
+// CopyNonGround returns the ref as-is if all parts are ground. Otherwise
+// returns a new ref with deep copies of the non-ground parts and shallow copies
+// of the ground parts. This is a *much* cheaper operation than Copy for operations
+// that only intend to modify (e.g. plug) the non-ground parts. The head element of
+// the ref is always shallow copied.
+func (ref Ref) CopyNonGround() Ref {
+	if ref.IsGround() {
+		return ref
+	}
+
+	cpy := make(Ref, len(ref))
+	cpy[0] = ref[0]
+
+	for i := 1; i < len(ref); i++ {
+		if ref[i].Value.IsGround() {
+			cpy[i] = ref[i]
+		} else {
+			cpy[i] = ref[i].Copy()
+		}
+	}
```

Comment:
Hmm how about
```suggestion
	for j := range len(ref)-1 {
		i := j+1
		if ref[i].Value.IsGround() {
			cpy[i] = ref[i]
		} else {
			cpy[i] = ref[i].Copy()
		}
	}
```
hmm no. 😁